### PR TITLE
[`core`] fix DP issue

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

The mini batching PR introduced a small bug that led to DP crash using `accelerate` as we tried to all reduce tensors that did not had the same shape. 
The tensors that did not had the same shape were the logprobs tensors, that derive from the inputs. 

The fix is to add a check, that padds correctly the model inputs, and corresponding attention masks, and fills them with the correct value using `accelerator.pad_across_processes`

cc @lvwerra 

run for gpt2 sentiment with DP=2: https://wandb.ai/distill-bloom/trl/runs/p7pzc7yk?workspace=user-younesbelkada
run for t5 sentiment with DP=2: https://wandb.ai/distill-bloom/trl/runs/nb9as16x?workspace=user-younesbelkada

### CODE REVIEW

**Review by lvwerra — COMMENTED** [review]
Left two comments, generally looks good.

File: trl/trainer/ppo_trainer.py [hunk 1]
```diff
@@ -441,6 +441,38 @@ def step(
...
         model_inputs = self.prepare_model_inputs(queries, responses)
+
+        if self.is_distributed:
+            pad_first = self.tokenizer.padding_side == "left"
+
+            if not self.is_encoder_decoder:
+                model_inputs["input_ids"] = self.accelerator.pad_across_processes(
+                    model_inputs["input_ids"], dim=1, pad_index=self.tokenizer.pad_token_id, pad_first=pad_first
+                )
+                model_inputs["labels"] = self.accelerator.pad_across_processes(
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply trl/trainer/ppo_trainer.py]: thanks, should be fixed in [e86d29a](https://github.com/lvwerra/trl/pull/222/commits/e86d29aac583176611619448286d7810663fb222)

File: trl/trainer/ppo_trainer.py [hunk 2]
```diff
@@ -441,6 +441,38 @@ def step(
...
+                    model_inputs["attention_mask"], dim=1, pad_index=0, pad_first=pad_first
+                )
+            else:
+                model_inputs["decoder_input_ids"] = self.accelerator.pad_across_processes(
+                    model_inputs["decoder_input_ids"],
+                    dim=1,
+                    pad_index=self.tokenizer.pad_token_id,
+                    pad_first=pad_first,
+                )
+                model_inputs["input_ids"] = self.accelerator.pad_across_processes(
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply trl/trainer/ppo_trainer.py]: Should be fixed in [e86d29a](https://github.com/lvwerra/trl/pull/222/commits/e86d29aac583176611619448286d7810663fb222)

File: trl/trainer/ppo_trainer.py [hunk 3]
```diff
@@ -445,30 +445,18 @@ def step(
...
-            else:
+            model_inputs["input_ids"] = self.accelerator.pad_across_processes(
+                model_inputs["input_ids"], dim=1, pad_index=self.tokenizer.pad_token_id, pad_first=pad_first
+            )
+            model_inputs["attention_mask"] = self.accelerator.pad_across_processes(
+                model_inputs["attention_mask"], dim=1, pad_index=0, pad_first=pad_first
+            )
+            if self.is_encoder_decoder:
                 model_inputs["decoder_input_ids"] = self.accelerator.pad_across_processes(
-                    model_inputs["decoder_input_ids"],
```
Comment (lvwerra) [inline trl/trainer/ppo_trainer.py]: I think this line should have stayed :)

**Review by lvwerra — APPROVED** [review]

File: trl/trainer/ppo_trainer.py (hunk 3 above)
Comment (younesbelkada) [reply trl/trainer/ppo_trainer.py]: yes haha
should be fixed in [3e441da](https://github.com/lvwerra/trl/pull/222/commits/3e441da08340c20b45a547ceb410964928b6690c)

**Comment by younesbelkada** [comment]
A CI test (`test_ppo_step_with_no_ref_sgd_lr_scheduler`) is surprisingly failing but completely unrelated to this PR, I thought it was an issue with PT 2.0 but still cannot reproduce locally. Will merge like this and investigate later
